### ISSUE TITLE
Implement "soft" stop

### ISSUE THREAD

**Issue by vicnevicne** [kokorin/Jaffree#81]
As discussed in #79 , it would be interesting to be able to send ffmpeg an exit request, which basically consists of emulating a "q" keystroke.
This issue is aimed at discussing implementation of that feature.

**Comment by vicnevicne** [comment]
I admit I'm having a hard time integrating with stdInWriter, so I first wanted to check feasability of the idea. Here is a successful (non-Jaffree) proof of concept:
```
package examples.ffmpeg;

import java.io.*;

public class SoftStop {
    public static void main(String[] args) throws IOException, InterruptedException {
        ProcessBuilder builder = new ProcessBuilder("ffmpeg", "-f", "lavfi", "-i", "testsrc=duration=3600:size=640x480:rate=30", "-y", "dummy.mp4");
        final Process process = builder.start();
        logOnConsole(process.getInputStream(), "OUT - ");
        logOnConsole(process.getErrorStream(), "ERR - ");

        Thread.sleep(5000);

        final OutputStream outputStream = process.getOutputStream();
        outputStream.write("q".getBytes());
        outputStream.flush();

        process.waitFor();
    }

    private static void logOnConsole(final InputStream inputStream, final String linePrefix) {
        new Thread(new Runnable() {
            @Override
            public void run() {
                try {
                    BufferedReader is = new BufferedReader(new InputStreamReader(inputStream));
                    String line;
                    while((line = is.readLine()) != null) {
                        System.out.println(linePrefix + line);
                    }
                }
                catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }).start();
    }
}
```

**Comment by vicnevicne** [comment]
Now my attempts at injecting a custom `StdWriter` in FFmpeg were unsuccessful until now, and I admit I'm a bit puzzled by the prototype in the interface, namely
```
void write(OutputStream stdIn);
```
I admit I am in no way a specialist of stream-oriented apps, but I would have expected the interface to expose a `write(int)`, `write(char)`, `write(String)` or the like for external processes to inject chars into the `OutputStream` exposed by the `Process`.
As (unless I'm mistaken) that `StdWriter` interface currently has no implementation in Jaffree, I tried to understand the way it is meant to be used. 
I found that the `write(OutputStream)` method is currently called upon process launch, with `process.getOutputStream()` as argument, so I guess `write` plays the role of a setter, so that a concrete implementation could store it and write to it later on. Is that correct ? 
If so, I would be tempted to rename that method (or to deprecate it and expose it) under the name 'setOutputStream()'. What do you think ?

Now I tried to write a StdWriterImpl that does just that, as follows:
```
private static class ControllerWriter implements StdWriter {
    BufferedWriter writer;

    @Override
    public void write(OutputStream stdIn) {
        writer = new BufferedWriter(new OutputStreamWriter(stdIn));
    }

    public void exit() {
        try {
            writer.write('q');
            writer.flush();
        } catch (IOException e) {
            throw new RuntimeException("Failed to write to stdin", e);
        }
    }
}
```
But it fails with
```
Exception in thread "main" java.lang.RuntimeException: Failed to write to stdin
	at examples.ffmpeg.Stop$ControllerWriter.exit(Stop.java:133)
	at examples.ffmpeg.Stop.stopWithQuitRequest(Stop.java:88)
	at examples.ffmpeg.Stop.main(Stop.java:105)
Caused by: java.io.IOException: Stream Closed
	at java.io.FileOutputStream.writeBytes(Native Method)
	at java.io.FileOutputStream.write(FileOutputStream.java:326)
	at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
```
And that is absolutely normal because the `ProcessHandler` calls the `write` method in a try-with-resource block, as follows:
```
    try (OutputStream outputStream = process.getOutputStream()) {
        stdInWriter.write(outputStream);
    } catch (Exception e) {
        throw new RuntimeException("Error while writing to Process", e);
    }
```
So basically, as soon a the `Thread` is started, the process's outputstream is retrieved, sent to the StdWriter... and immediately closed :-(

I would be tempted to think it is a mistake, but the comment above seems to indicate it is done on purpose:
```
    // Explicitly close stdIn to notify process, that there will be no more data
```
I really don't understand the goal behind that "process notification".
Is that to save resources ? And if so why not call `process.getOutputStream().close()` ?

I'm willing to change that behaviour, but I fear that the code and comment are there for a reason. 

Could you shed some light on this ?

Kind regards, Vicne

**Comment by kokorin** [comment]
```
        if (stdInWriter != null) {
            executor.execute("StdIn", new Runnable() {
                @Override
                public void run() {
                    // Explicitly close stdIn to notify process, that there will be no more data
                    try (OutputStream outputStream = process.getOutputStream()) {
                        stdInWriter.write(outputStream);
                    } catch (Exception e) {
                        throw new RuntimeException("Error while writing to Process", e);
                    }
                }
            });
        }
```

This code is a remnant from the times when Jaffree accepted media with StdIn and wrote as StdOut. 

Now it's not needed as Jaffree can accept InputStream as PipeInput and can write to OutputStream with PipeOutput. It appeared that TCP interaction is much faster, so there is no need in passing data via StdIn or getting it via StdOut.

Thank you for providing working example with graceful stopping of ffmpeg. I will try to adapt existing code to allow such behaviour.

**Comment by kokorin** [comment]
I have drafted a new branch with possible fix: https://github.com/kokorin/Jaffree/tree/GRACEFUL_STOP

Sorry, but I have no more time now. Check it please.

**Comment by vicnevicne** [comment]
Thanks for the explanations, makes much more sense now. 

And thanks for taking the time to wrap up the test branch.  I gave it a go by modifying the "Stop" example (by the way, I noticed a copy/paste typo: `stopWithInterruption()`is [called a second time](https://github.com/kokorin/Jaffree/blob/cdd1f09e3d202896fa96bb696dc9a9e308c283cc/src/test/java/examples/ffmpeg/Stop.java#L78) instead of `stopWithFutureCancellation()`.)

I added a 4th stop method which is similar to the third except I pass 'false' to `cancel()`:
```
    public static void stopWithQuitRequest(final FFmpeg ffmpeg) throws Exception {
        Future<FFmpegResult> future = ffmpeg.executeAsync();

        Thread.sleep(5_000);
        // We pass mayInterruptIfRunning false to just send a 'quit' request to ffmpeg
        future.cancel(false);

        Thread.sleep(1_000);
        System.out.println(future.get());
    }
```
**Good news:** it works :-). 
If I write to an actual file instead of `NullOutput`, the FFmpeg process exits but I'm getting a valid file.

**Bad news:** The 'quit' request requires some time for the ffmpeg process to end, which leads to 2 small issues:
1. I'm afraid the `super.cancel()` follows so closely the quit request that it might try to interrupt the ffmpeg process while it is still shutting down. The proof ffmpeg is still running is that `super.cancel()` returns true (according to the spec, it returns false if the task has already completed normally).
2. More importantly, the `future.get()` that comes next in the `Stop` example throws an exception because the process is still running:
```
Exception in thread "main" java.util.concurrent.CancellationException
	at java.util.concurrent.FutureTask.report(FutureTask.java:121)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at examples.ffmpeg.Stop.stopWithQuitRequest(Stop.java:78)
	at examples.ffmpeg.Stop.main(Stop.java:92)
```
As a "quick and dirty" fix, I tried adding a delay after the quit request, like so:
```
final FutureTask<FFmpegResult> result = new FutureTask<FFmpegResult>(callable) {
    @Override
    public boolean cancel(boolean mayInterruptIfRunning) {
        if (!mayInterruptIfRunning) {
            stopStdWriter.stop();
            // #### Wait for 1 sec. #### 
            try { Thread.sleep(1000); } catch (InterruptedException e) {}
        }
        return super.cancel(mayInterruptIfRunning);
    }
};
```
That does the trick, as `super.cancel()` now returns false as expected, and the `future.get()` returns a valid object.

Now that's only a dirty trick, because 1 second might be wasting user's time, while on an overloaded system, I could imagine it would not be sufficient to flush a file to a slow disk.
Plus, there's a [delay of up to 100ms](https://github.com/kokorin/Jaffree/blob/f2cc1fb02e48e2a9e973e9f84ed74e9099d4ae3a/src/main/java/com/github/kokorin/jaffree/ffmpeg/StopStdWriter.java#L36) in the `StopStdWriter` before the request is even transmitted to the FFmpeg process.
To avoid that loop, I would be tempted to simplify the StopStdWriter as follows:
```
public class StopStdWriter implements StdWriter {
    private static final Logger LOGGER = LoggerFactory.getLogger(StopStdWriter.class);
    private OutputStream stdIn;

    @Override
    public void write(OutputStream stdIn) {
        this.stdIn = stdIn;
    }

    public void stop() {
        try {
            stdIn.write('q');
        }
        catch (IOException e) {
            LOGGER.info("Ignoring {}: {}", e.getClass(), e.getMessage());
        }
    }
}
```
I don't see an issue in keeping a reference to stdIn in a field, but maybe I'm missing something ?

As for the actual shutdown delay of FFmpeg, I guess the cleanest way would be to call `process.waitFor()` or equivalent, and only return when the process has actually ended. It should be wrapped in a "Timeout thread" though, to forcibly kill FFmpeg and return after a reasonable amount of time (10 sec ?) in case it is hung, but that's failrly easy.

The hardest part is to get a reference to the Process inside the StopStdWriter...
Passing a `Process` parameter to `write()` and storing it in a field instead of the `OutputStream` would suffice as the `stop()` method could call `process.getOutputStream().write('q')` before `process.waitFor()`, but from an architecture point of view, this is rather ugly :-(
What do you think ?

KR, Vicne

**Comment by kokorin** [comment]
Please check  again https://github.com/kokorin/Jaffree/tree/GRACEFUL_STOP

I have removed StdIn closing in ProcessHandler, added flush and do not call super `Future.cancel()` if ffmpeg should be stopped gracefully.

This solution clearly violates Liskov substitution principle because now `Future.get()` doesn't throw `CancellationException` when `Future.cancel()` is called. But it works.

I think it's possible to replace `FFmpeg.executeAsync()` return type to something like `FFmpegResultFuture` or `FFmpegProcess` which will support `cancel(boolean)` method for backward compatibility.

Not sure yet, I need some time to think about it.

> The hardest part is to get a reference to the Process inside the StopStdWriter

That's tricky and error-prone idea, StdWriter should be responsible for only one thing: writing, without any control on system process.

> if you rename "StopStdListener" to "StopProcessHelper", all of a sudden it's much cleaner ;-)

Did you mean `StopStdWriter`?

> if you want, I can try to wrap up something based on that idea of a separate "StopProcessHelper", but I don't want to step on you toes.

Honestly it's always good to hear someelse's opinion.

**Comment by kokorin** [comment]
I have added `FFmpegResultFuture`. I delegates all methods to `Futures<FFmpegResult` except for `cancel(boolean)`.

Not still sure if this should be implemented in that way. Also notice that I replaced `Thread.sleep(100)` in `StopStdWriter` with `wait()` and `notifyAll()`. This eliminates 100 ms delay between request for stop and actual stop.

**Comment by vicnevicne** [comment]
Sorry, I didn't have time today to check your modifications. I'll do that tomorrow. 
Sounds promising. Thanks very much already !

**Comment by vicnevicne** [comment]
Hi Denis. 

I'd be glad to integrate the slight changes proposed above in your code, but I don't know if I should work on your branch, or create a separate branch in my fork and wrap up a CR - sorry if this is a newbie question, but I'm not that fluent in Github as you see. How do you want to proceed ? 

On the other hand, maybe you have thought of a better/cleaner way to implement that feature ?

As I said, I think one key point could be to completely separate the "quit request" (q key) from the notion of stopping a running progress. Maybe it would help us to not focus on "q" in particular but to see the big picture of all keystrokes that FFmpeg recognizes when running. In my version, pressing '?' when FFmpeg runs returns:
```
key    function
?      show this help
+      increase verbosity
-      decrease verbosity
c      Send command to first matching filter supporting it
C      Send/Queue command to all matching filters
D      cycle through available debug modes
h      dump packets/hex press to cycle through the 3 states
q      quit
s      Show QP histogram
```
Don't get me wrong, I don't mean we should _implement_ any of them except 'q', or not in a first phase at least. I just mean that seeing 'q' just as "one of the many ways to interact with a running ffmpeg instance" would shed another light on the topic.

In that sense, keeping it as a separate StdInWriter would make much sense.

Just my 2 cents.

**Comment by kokorin** [comment]
Hello @vicnevicne ,

Honestly I couldn't find use cases for graceful ffmpeg stop before your request. And you are the first in almost 3 years.
Probably the same will be with extra keystrokes. I'm always stick to Occam's Razor principle (old good KISS), so I see no need ti support of extra keystrokes except `q` at the moment.

On the weekend I will try to spend some time on reviewing this issue & PR and, merge to master and probably publish new release to Maven central.

**Comment by vicnevicne** [comment]
That would be great !
And as I said, I never intended to support anything else than "q", just seeing "q" as a normal termination, like if a transcode operation was finished, not as a "kill" command.
But as I said the implementation on the branch suits my needs 100%, so I really trust you on the best/simplest/cleanest way to integrate it.

**Comment by kokorin** [comment]
Implemented in PR #85
